client2/arq.go

```diff
@@ -0,0 +1,241 @@
+
+	sphinxComposerSender SphinxComposerSender
+}
+
+// NewARQ creates a new ARQ.
+func NewARQ(sphinxComposerSender SphinxComposerSender, logbackend io.Writer) *ARQ {
+	log := log.NewWithOptions(logbackend, log.Options{
+		ReportTimestamp: true,
+		Level:           log.DebugLevel,
+		Prefix:          "_ARQ_",
+	})
+	log.Info("NewARQ")
+	return &ARQ{
+		log:                  log,
+		gcSurbIDCh:           make(chan *[sConstants.SURBIDLength]byte),
+		surbIDMap:            make(map[[sConstants.SURBIDLength]byte]*ARQMessage),
+		sphinxComposerSender: sphinxComposerSender,
+	}
+}
+
+// Start starts the ARQ worker thread. You MUST start before using.
+func (a *ARQ) Start() {
+	a.log.Info("Start")
+	a.timerQueue = NewTimerQueue(func(rawSurbID interface{}) {
+		a.log.Info("TimerQueue callback!")
+		surbID, ok := rawSurbID.(*[sConstants.SURBIDLength]byte)
+		if !ok {
+			panic("wtf, failed type assertion!")
+		}
+		a.resend(surbID)
+	})
+	a.timerQueue.Start()
+}
+
+// Stop stops the ARQ's timer queue worker thread.
+func (a *ARQ) Stop() {
+	a.log.Info("Stop")
+	a.timerQueue.Halt()
+	a.timerQueue.Wait()
+}
+
+func (a *ARQ) resend(surbID *[sConstants.SURBIDLength]byte) {
+	a.log.Info("resend")
+
+	a.lock.Lock()
+	message, ok := a.surbIDMap[*surbID]
+	if ok {
+		delete(a.surbIDMap, *surbID)
+		newsurbID := &[sConstants.SURBIDLength]byte{}
+		_, err := rand.Reader.Read(newsurbID[:])
+		if err != nil {
+			panic(err)
+		}
+
+		pkt, k, rtt, err := a.sphinxComposerSender.ComposeSphinxPacket(&Request{
+			WithSURB:          true,
+			SURBID:            newsurbID,
+			DestinationIdHash: message.DestinationIdHash,
+			RecipientQueueID:  message.RecipientQueueID,
+			Payload:           message.Payload,
+			IsSendOp:          true,
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		message.SURBID = newsurbID
+		message.Key = k
+		message.ReplyETA = rtt
+		message.SentAt = time.Now()
+		message.Retransmissions += 1
+
+		a.surbIDMap[*newsurbID] = message
+		priority := uint64(message.SentAt.Add(message.ReplyETA).Add(RoundTripTimeSlop).UnixNano())
+		a.timerQueue.Push(priority, surbID)
+
+		err = a.sphinxComposerSender.SendSphinxPacket(pkt)
+		if err != nil {
+			a.log.Errorf("gc sphinx composer failure: %s", err.Error())
+		}
+
+	} else {
+		a.log.Error("gc SURB ID not found")
+	}
+	a.lock.Unlock()
+}
+
+// Has checks if a given SURB ID exists.
+func (a *ARQ) Has(surbID *[sConstants.SURBIDLength]byte) bool {
+	a.log.Info("Has")
+
+	a.lock.RLock()
+	_, ok := a.surbIDMap[*surbID]
+	a.lock.RUnlock()
+	return ok
+}
+
+// HandleACK removes the map entry for the given SURB ID AND returns
+// the APP ID and SURB Key so that the reply and be decrypted and routed
+// to the correct application.
+func (a *ARQ) HandleAck(surbID *[sConstants.SURBIDLength]byte) (*replyDescriptor, error) {
+	a.lock.Lock()
+
+	m, ok := a.surbIDMap[*surbID]
+	if !ok {
+		a.log.Error("failed to find SURB ID in ARQ map")
+		return nil, errors.New("failed to find SURB ID in ARQ map")
```

Comment:
need to relinquish the lock here or we deadlock.
probably easiest to just `defer a.lock.Unlock()` at the top here

Reply:
fixed